internal/model/target.go

```diff
@@ -82,3 +86,55 @@ func DedupeTargetIDs(ids []TargetID) []TargetID {
 	}
 	return result
 }
+
+// Map all the targets by their target ID.
+func MakeTargetMap(targets []TargetSpec) map[TargetID]TargetSpec {
+	result := make(map[TargetID]TargetSpec, len(targets))
+	for _, target := range targets {
+		result[target.ID()] = target
+	}
+	return result
+}
+
+// Create a topologically sorted list of targets. Returns an error
+// if the targets can't be topologically sorted. (e.g., there's a cycle).
+func TopologicalSort(targets []TargetSpec) ([]TargetSpec, error) {
+	targetMap := MakeTargetMap(targets)
+	result := make([]TargetSpec, 0, len(targets))
+	inResult := make(map[TargetID]bool, len(targets))
+	searching := make(map[TargetID]bool, len(targets))
+
+	var ensureInResult func(id TargetID) error
```

Comment:
Why are these on two separate lines?

Reply:
this is a common go faq
https://stackoverflow.com/questions/28099441/define-a-recursive-function-within-a-function-in-go

Reply:
Ah! TIL thank you